Refactor Slider (#3296)
jpmorganchase/salt-ds: site/docs/components/slider/examples.mdx

## With form field

To display a label alongside a slider, use FormField and FormFieldLabel. For more information, refer to the [form field documentation.](/salt/components/form-field/examples)

### Best practices

We recommend using labels with slider to provide context and aid the user in understanding the value being selected.

</LivePreview>

<LivePreview componentName="slider" exampleName="WithInput">

## With input

You can display inputs with either a single or range slider. Inputs are used to enter a numeric value from the predefined range, enabling the thumb to jump to the specified value.

"Out of range" changes to the value (i.e. setting the input to a value above or below the slider's `min` and `max` values) will be rounded to the nearest valid value for the slider. When using a range slider, the upper and lower values cannot overlap (i.e. the lower value cannot be set to a value higher than the upper value). If the input value is set to a value that would cause an overlap, the range value will be set to the nearest valid value.

### Best practices

- For single slider, display one input to the left of the track.

> - When using inputs, ensure the slider is in "controlled" mode (by explicitly setting the `value` prop).

Reviewer: This doesn't seem like a best practice?

> - Use the slider's `onChange` callback to update the input values when the slider is interacted with.

Reviewer: ```suggestion
-  Update the relevant input's value as the Slider's value changes.
```

- Validate your input value changes and update the `validationStatus` prop on any inputs accordingly.

> - When using the slider alongside an input use the default `marks` setting (do not set `marks` to `"bottom"` or `"all"`).

Reviewer: We should probably mention what the default is:

```suggestion
- When using the slider alongside an input use the default `"inline"` `marks` setting (do not set `marks` to `"bottom"` or `"all"`).
```

---
title:
  $ref: ./#/title
layout: DetailComponent
sidebar:
  exclude: true
data:
  $ref: ./#/data
---

<LivePreviewControls>

> <LivePreview componentName="slider" exampleName="Single">

Reviewer: Theses examples will need display name to match the displayed title.

> ## Single Slider

Reviewer: ```suggestion
## Single slider
```

> Selection of a single value from a horizontally oriented slider track. Values can be displayed either inline or below the track.

Reviewer: If we call them "marks", we shouldn't also refer to them as "values"

> Selection of a single value from a horizontally oriented slider track. Values can be displayed either inline or below the track.

Reviewer 2: Agreed. I've taken these out

A tooltip is displayed when the thumb is dragged to ensure that incremental values remain visible when the number increments are not displayed along the track (i.e., with vertical label position).

</LivePreview>

<LivePreview componentName="slider" exampleName="Range">

> ## Range Slider

Reviewer: ```suggestion
## Range slider
```

> Selection of a range of values from a horizontally oriented slider track. This determines a minimum and maximum range. Values can be displayed either inline or below the track.

Reviewer: "This" is undefined. It sounds like a joke haha but I'm not sure what here determines a minimum and maximum range.

### Best practices

> The default minimum range is set to 1 step, therefore thumbs cannot pass or overlap.

Reviewer: This might need to be re-worded so that the best practice is guidance for a consumer.

</LivePreview>

<LivePreview componentName="slider" exampleName="CustomStep">

> ## Custom Step

Reviewer: ```suggestion
## Custom step
```

> By default, the slider step increment is set to 1. This can be customised by configuring the ‘step’ prop. We recommend using an increment most suitable to your use case by using a step relative to the granularity of the scale. This prevents slider selections from becoming tedious and difficult to manage.

Reviewer: ```suggestion
By default, the slider step increment is set to 1. This can be customized using the ‘step’ prop. We recommend using an increment most suitable to your use case by using a step relative to the granularity of the scale. This prevents slider selections from becoming tedious and difficult to manage.
```

Custom step values can be set to any number, including decimal values up to two decimal places. If a decimal step value is provided, the marks (if shown) will also be displayed as decimal values.

> If a custom step value is provided that is not a multiple of the range, the slider will round the maximum selectable value to the largest multiple of the step that is less than the maximum value. **Note:** This will render the final portion of the slider track unusable (the portion between the rounded maximum value and the actual maximum value). Step calculations are made from `0`, so the same rounding logic applies to negative values, but in the opposite direction.

Reviewer: Do we show this behaviour anywhere? I couldn't see it in the example

Reviewer 2: Added an example (and extra labels to clarify the other custom step examples)

Reviewer: I might be missing it but the example still don't show the final portion of the track being unusable? Also not sure if this is tested in the tests anywhere

</LivePreview>

<LivePreview componentName="slider" exampleName="SingleFormField">

> ## Single Form Field

Reviewer: ```suggestion
## Single form field
```

> Wrap single slider in a form field to display a left-aligned or top label.

Reviewer: The grammar seems odd here, might need a tweak "left-aligned or top label" left and top are both alignments of the label.

Reviewer 2: Reworded entirely 👍

For more information, refer to the [form field documentation.](/salt/components/form-field/examples)

### Best practices

We recommend using labels with slider to provide context and aid the user in understanding the value being selected.

</LivePreview>

<LivePreview componentName="slider" exampleName="RangeFormField">

> ## Range Form Field

Reviewer: ```suggestion
## Range form field
```

Wrap range slider in a form field to display a left-aligned or top label.

For more information, refer to the [form field documentation.](/salt/components/form-field/examples)

### Best practices

> We recommend using labels with slider to provide context and aid the user in understanding the value being selected.

Reviewer: Should values be pluralised here?

Reviewer 2: Have reworded and simplified here. Combined "Single Form Field" and "Range Form Field" into a new "With Form Field" example. There's no real difference to using single/range in a form field, so makes more sense for there to be only one example.

Reviewer: Is there a reason why we need to use a different copy to the other form controls?

</LivePreview>

<LivePreview componentName="slider" exampleName="WithInput">

> ## With Input

Reviewer: ```suggestion
## With input
```

> You can display inputs with either a single or range slider. Inputs are used to enter a numeric value from the predefined range, enabling the thumb to jump to the specified value.

Reviewer: This feels backwards, with it focusing on inputs first. Also this doesn't mention why inputs are useful

Reviewer 2: See rewording mentioned in earlier comment.

Reviewer: I still feel like we can do more to explain why you would want to supplement a slider with an input

> The thumb and input are in sync. An update to either causes the other to update. E.g., if you use cursor click or drag along the slider track, the input must display the change in value live.

Reviewer: if you use "cursor click", I feel like this needs to be re-worded

Reviewer: ```suggestion
The thumb and input are in sync. An update to either causes the other to update. E.g., if you use click or drag along the slider track, the input must display the change in value live.
```

Reviewer 2: See rewording mentioned in earlier comment.

### Best practices

For single slider, display one input to the left of the track.

> Only use the inline variant of slider alongside an input. This ensures the size-base alignment of the overall component layout is maintained.

Reviewer: This should be a list

> Only use the inline variant of slider alongside an input. This ensures the size-base alignment of the overall component layout is maintained.

Reviewer: What is the "inline variant" of slider?

Reviewer 2: I've reworded this whole section. Should be clearer now in conjunction with fleshed-out "Best practices"
